pkg/client/simple/api/clientset.go

```diff
@@ -0,0 +1,198 @@
+	"k8s.io/kops/util/pkg/vfs"
+	"k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/api/errors"
+)
+
+// RESTClientset is an implementation of clientset that uses a "real" generated REST client
+type RESTClientset struct {
+	BaseURL    *url.URL
+	KopsClient kopsinternalversion.KopsInterface
+}
+
+// GetCluster implements the GetCluster method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) GetCluster(name string) (*kops.Cluster, error) {
+	namespace := restNamespaceForClusterName(name)
+	return c.KopsClient.Clusters(namespace).Get(name, metav1.GetOptions{})
+}
+
+// CreateCluster implements the CreateCluster method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) CreateCluster(cluster *kops.Cluster) (*kops.Cluster, error) {
+	namespace := restNamespaceForClusterName(cluster.Name)
+	return c.KopsClient.Clusters(namespace).Create(cluster)
+}
+
+// UpdateCluster implements the UpdateCluster method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) UpdateCluster(cluster *kops.Cluster, status *kops.ClusterStatus) (*kops.Cluster, error) {
+	glog.Warningf("validating cluster update client side; needs to move to server")
+	old, err := c.GetCluster(cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+	if err := validation.ValidateClusterUpdate(cluster, status, old).ToAggregate(); err != nil {
+		return nil, err
+	}
+
+	namespace := restNamespaceForClusterName(cluster.Name)
+	return c.KopsClient.Clusters(namespace).Update(cluster)
+}
+
+// ConfigBaseFor implements the ConfigBaseFor method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) ConfigBaseFor(cluster *kops.Cluster) (vfs.Path, error) {
+	if cluster.Spec.ConfigBase != "" {
+		return vfs.Context.BuildVfsPath(cluster.Spec.ConfigBase)
+	}
+	// URL for clusters looks like  https://<server>/apis/kops/v1alpha2/namespaces/<cluster>/clusters/<cluster>
+	// We probably want to add a subresource for full resources
+	return vfs.Context.BuildVfsPath(c.BaseURL.String())
+}
+
+// ListClusters implements the ListClusters method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) ListClusters(options metav1.ListOptions) (*kops.ClusterList, error) {
+	return c.KopsClient.Clusters(metav1.NamespaceAll).List(options)
+}
+
+// InstanceGroupsFor implements the InstanceGroupsFor method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) InstanceGroupsFor(cluster *kops.Cluster) kopsinternalversion.InstanceGroupInterface {
+	namespace := restNamespaceForClusterName(cluster.Name)
+	return c.KopsClient.InstanceGroups(namespace)
+}
+
+// FederationsFor implements the FederationsFor method of Clientset for a kubernetes-API state store
+func (c *RESTClientset) FederationsFor(federation *kops.Federation) kopsinternalversion.FederationInterface {
+	// Unsure if this should be namespaced or not - probably, so that we can RBAC it...
+	panic("Federations are currently not supported by the server API")
```

Comment:
`panic` ouch.  We do not do that in many places. You want the server to restart?

Reply:
It's a panic because there's no error code here, and because we don't expect people to be using Federations.   I'd say panic is appropriate - our code should be crash-safe.